File: crates/onetaskgraph/tests/e2e/document_store.py
"""A file-backed document source, spoken to over `docs/plugin-protocol.md`.

It exists for one reason the suite could not otherwise satisfy. A copy is only proven by
reading the destination back *afterwards*, and every destination this build has is either
a folder of Markdown — which declares it has no documents — or the in-memory source, whose
work dies with the process that held it. One command-line invocation is one process, so a
document copy driven the way a user drives it had nothing left to look at. This keeps its
documents in a JSON file, so what one invocation writes the next one reads.

It is written in Python on purpose, and not only to keep a fixture out of the shipped
binary. The stdio seam's whole claim is that a plugin can be written in another language
against the protocol document alone; this peer shares not one line with the engine's own
half, so the journeys that drive it test that claim rather than restating the engine's
implementation of it back to itself. `python3` is already what every guard under
`workspace:lint` runs on all three platforms, so it costs the gate no new dependency.

Its settings, handed over in the `initialize` request (§3), are
`{"store": <path>, "documents": "native"|"unsupported", "log": <path>}` — the last two
optional. `documents` lets a journey configure a peer that says it has none, and `log`
appends the name of every method this source is asked for, which is how a journey proves a
refusal happened *before* anything was read.
"""

# llmlint: ignore-file[modern_domain_modeling] This peer is a transcription of
# docs/plugin-protocol.md, and the protocol's own types are JSON objects. It is spawned
# with a cleared environment (§3.1) by whatever `python3` or `python` the host provides, so
# it may import nothing outside the standard library and cannot assume a version that has
# the typing this rule asks for. A stdlib model layer would also be erased at runtime,
# adding no check that `parameter`, `read_store` and `initialize` below do not already
# make — while costing the file the property the journeys rely on: that a plugin can be
# written from the protocol document alone, sharing no type with the engine's own half.

import json
import os
import sys

KIND = "document-store"
PROTOCOL_VERSION = 2

# The largest page this source will serve, which it also declares at the handshake.
MAX_PAGE_SIZE = 50


def refused(message):
    """The contract's `SourceError::Refused`, in its serialized shape."""
    return {"kind": "refused", "message": message}


def malformed(message):
    """The contract's `SourceError::Malformed`."""
    return {"kind": "malformed", "message": message}


def read_store(path):
    """The documents on disk, treating a file that is not there yet as an empty store.

    A missing file is the ordinary first-run state — a destination nothing has been copied
    into yet — rather than a failure. A file that is there and is not a store is a
    different thing, and says so rather than raising a Python error into the pipe.
    """
    try:
        with open(path, encoding="utf-8") as handle:
            held = json.load(handle)
    except FileNotFoundError:
        return []
    except ValueError:
        raise Refusal(malformed("%s is not JSON, so it is not this source's store" % path))
    documents = held.get("documents", []) if isinstance(held, dict) else None
    if not isinstance(documents, list):
        raise Refusal(malformed('%s is not a store: expected {"documents": [...]}' % path))
    return documents


def write_store(path, documents):
    """Write the documents back, creating the directory they live in."""
    parent = os.path.dirname(path)
    if parent:
        os.makedirs(parent, exist_ok=True)
    with open(path, "w", encoding="utf-8") as handle:
        json.dump({"documents": documents}, handle, indent=2)


def note(settings, method):
    """Record that this source was asked for `method`, when its settings ask for a record."""
    path = settings.get("log")
    if path is None:
        return
    with open(path, "a", encoding="utf-8") as handle:
        handle.write(method + "\n")


def capabilities(settings):
    """What this source declares once, at the handshake.

    No projects of its own, and it says so: the engine then reports that predicate
    unavailable for this source rather than reading an empty page as a narrowed answer.
    """
    return {
        "projects": "unsupported",
        "documents": settings.get("documents", "native"),
        "orphan_tasks": "native",
        "filter_by_label": "native",
        "filter_by_status": "native",
        "search_title": "native",
        "search_content": "native",
        "task_dependencies": "both-directions",
        "project_dependencies": "both-directions",
        "max_page_size": MAX_PAGE_SIZE,
    }


def has_documents(settings):
    """Whether this source has documents at all, as its settings declared."""
    return settings.get("documents", "native") == "native"


def matches_text(document, query):
    """Whether one document matches a free-text query, over the fields it names."""
    if query is None:
        return True
    terms = query["terms"].lower()
    in_title = terms in (document.get("title") or "").lower()
    in_content = terms in (document.get("content") or "").lower()
    fields = query["fields"]
    if fields == "title":
        return in_title
    if fields == "content":
        return in_content
    return in_title or in_content


def survives(document, query):
    """Whether one document survives every predicate a query carries."""
    held = [label["name"].lower() for label in document.get("labels") or []]
    labels = query.get("labels") or {}
    any_of = labels.get("any_of") or []
    if any_of and not any(name.lower() in held for name in any_of):
        return False
    if not all(name.lower() in held for name in labels.get("all_of") or []):
        return False
    if any(name.lower() in held for name in labels.get("none_of") or []):
        return False
    project = query.get("project", "any")
    if project == "orphans" and document.get("project") is not None:
        return False
    if isinstance(project, dict) and document.get("project") != project.get("is"):
        return False
    return matches_text(document, query.get("text"))


def paginate(items, page):
    """Slice `items` into the page asked for, refusing a cursor this source never issued."""
    cursor = page.get("cursor")
    start = 0
    if cursor is not None:
        if not cursor.isdigit() or int(cursor) >= len(items):
            raise Refusal(
                malformed(
                    "cursor %r was not issued by this source; it addresses no row of the %d "
                    "result(s) available" % (cursor, len(items))
                )
            )
        start = int(cursor)
    limit = page.get("limit", 0)
    if limit < 1:
        raise Refusal(
            {
                "kind": "config",
                "message": "a page limit of 0 is not a page; ask for at least 1 row",
            }
        )
    end = min(start + min(limit, MAX_PAGE_SIZE), len(items))
    return {
        "items": items[start:end],
        "next": str(end) if end < len(items) else None,
    }


def unused(documents, wanted):
    """`wanted` when nothing holds it, or the first `wanted-N` that is free.

    A destination decides its own ids, exactly as every other writable source does: the id
    an item was read under at its source is a suggestion.
    """
    taken = {document["id"] for document in documents}
    if wanted not in taken:
        return wanted
    attempt = 2
    while "%s-%d" % (wanted, attempt) in taken:
        attempt += 1
    return "%s-%d" % (wanted, attempt)


class Refusal(Exception):
    """A refusal this source answers one request with, carrying the contract's own shape."""

    def __init__(self, error):
        super().__init__(error["message"])
        self.error = error


def parameter(params, name, kind, method):
    """One request parameter, refused as `malformed` when it is absent or the wrong shape.

    A plugin answers a request it cannot parse with `malformed` (§5) rather than failing in
    a way the engine has to guess at, so this names the method and the parameter instead of
    raising a Python error into the pipe — where the engine would see only a dead peer.
    """
    value = params.get(name) if isinstance(params, dict) else None
    if not isinstance(value, kind):
        raise Refusal(
            malformed(
                "%s needs a %s parameter %r" % (method, kind.__name__, name)
            )
        )
    return value


def document_side(settings):
    """Refuse every document call when this source's settings say it has none.

    In the same words `documentless` gives every other document-free source, because a
    caller must not be able to tell which plugin refused apart from the kind it names.
    """
    if not has_documents(settings):
        raise Refusal(refused("the %s plugin has no documents" % KIND))


# llmlint: ignore-block[structural_pattern_matching] `match`/`case` is a syntax error
# before Python 3.10, and this peer is spawned with a cleared environment by whichever
# `python3` or `python` the host has on PATH — see `interpreter()` in document_store.rs.
# A syntax error there aborts the interpreter at import, which the engine sees as a plugin
# that closed its output rather than as a portability problem, so the whole document-copy
# round trip would report as broken on any host still shipping 3.9.
def dispatch(settings, method, params):
    """Answer one method against the store this source's settings name."""
    store = settings["store"]
    if method == "health":
        return {
            "reachable": True,
            "detail": "%d document(s) on disk" % len(read_store(store)),
        }
    # This source holds no tasks and no projects, and an empty page is the whole truth:
    # `projects: unsupported` at the handshake is what keeps that honest.
    if method in ("get_task", "get_project"):
        return {method.removeprefix("get_"): None}
    if method in (
        "query_tasks",
        "query_projects",
        "labels",
        "task_dependencies",
        "project_dependencies",
    ):
        return {"items": [], "next": None}
    if method == "get_document":
        document_side(settings)
        wanted = parameter(params, "id", str, method)
        found = [d for d in read_store(store) if d["id"] == wanted]
        return {"document": found[0] if found else None}
    if method == "query_documents":
        document_side(settings)
        query = parameter(params, "query", dict, method)
        kept = [d for d in read_store(store) if survives(d, query)]
        return paginate(kept, parameter(params, "page", dict, method))
    if method == "write_document":
        document_side(settings)
        write = parameter(params, "write", dict, method)
        documents = read_store(store)
        target = write.get("target")
        landing = dict(parameter(write, "item", dict, method))
        if target is None:
            landing["id"] = unused(documents, landing["id"])
            documents.append(landing)
        else:
            at = [i for i, d in enumerate(documents) if d["id"] == target]
            if not at:
                raise Refusal(
                    refused(
                        "%s names no document this source holds; next: copy with --recreate "
                        "to create one instead of updating" % target
                    )
                )
            landing["id"] = target
            documents[at[0]] = landing
        write_store(store, documents)
        return {"id": landing["id"]}
    if method == "delete_document":
        document_side(settings)
        unwanted = parameter(params, "id", str, method)
        documents = [d for d in read_store(store) if d["id"] != unwanted]
        write_store(store, documents)
        return {}
    if method in ("write_task", "write_project", "delete_task", "delete_project"):
        raise Refusal(refused("the %s plugin cannot be written" % KIND))
    raise Refusal(
        malformed("protocol version %d has no method called %r" % (PROTOCOL_VERSION, method))
    )


# llmlint: ignore-end[structural_pattern_matching]
def initialize(params):
    """The handshake (§3), which is also where this source learns where its store is."""
    version = params.get("protocol_version")
    if version != PROTOCOL_VERSION:
        raise Refusal(
            {
                "kind": "config",
                "message": "protocol version %s is not supported by this plugin; it speaks "
                "version %d" % (version, PROTOCOL_VERSION),
            }
        )
    settings = params.get("config") or {}
    # Every setting is read here, once, rather than trusted wherever it is later used: a
    # peer configured with a path that is not one, or a capability value the contract has
    # no member for, is a configuration mistake and says so at the handshake.
    if (
        not isinstance(settings, dict)
        or not isinstance(settings.get("store"), str)
        or settings.get("documents", "native") not in ("native", "unsupported")
        or not isinstance(settings.get("log", ""), str)
    ):
        raise Refusal(
            {
                "kind": "config",
                "message": 'this source\'s settings must be {"store": <path>, "documents": '
                '"native"|"unsupported", "log": <path>}, the last two optional',
            }
        )
    note(settings, "initialize")
    return settings, {
        "protocol_version": PROTOCOL_VERSION,
        "kind": KIND,
        "capabilities": capabilities(settings),
        "writes": "supported",
    }


def main():
    """Serve one connection until the engine closes its input."""
    settings = None
    for line in sys.stdin:
        if not line.strip():
            continue
        try:
            request = json.loads(line)
            identifier = request["id"]
        except (ValueError, KeyError, TypeError):
            print("%s: ignoring an unaddressed line" % KIND, file=sys.stderr)
            continue
        method = request.get("method", "")
        params = request.get("params") or {}
        try:
            if not isinstance(method, str) or not isinstance(params, dict):
                raise Refusal(malformed("a request names its method and carries an object"))
            if method == "initialize":
                settings, result = initialize(params)
            elif settings is None:
                raise Refusal(malformed("%s arrived before the handshake" % method))
            else:
                note(settings, method)
                result = dispatch(settings, method, params)
            answer = {"id": identifier, "result": result}
        except Refusal as refusal:
            answer = {"id": identifier, "error": refusal.error}
        sys.stdout.write(json.dumps(answer) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()

File: crates/onetaskgraph/tests/e2e/document_store.rs
//! The document copy round trip, against a destination that outlives the invocation.
//!
//! Every other copy journey reads its destination back through a later command, which is
//! only possible because a folder of Markdown is still there afterwards. Documents had no
//! such destination: `local-md` declares it holds none, and the in-memory source's work
//! dies with the process that held it — so a `document copy` typed at a shell could be
//! observed only through the report it printed, and "the document really landed carrying
//! every field" was proven one layer down, as a library call.
//!
//! This closes that. The destination is `onetaskgraph-document-store`, a peer that keeps
//! its documents in a JSON file and speaks `docs/plugin-protocol.md` over a real pipe to a
//! real second process — so the copy in one invocation is read back by the *next*
//! invocation, through the same command line a user types.

use std::path::{Path, PathBuf};
use std::process::Output;

use serde_json::{Value, json};

use crate::common::{Sandbox, stderr, stdout};
use crate::fixtures::{ROWS, Row, document, empty_folder, qualified};

/// The source every journey here copies out of.
const SOURCE: &str = "work";

/// The persistent destination they copy into.
const STORE: &str = "store";

fn run(sandbox: &Sandbox, arguments: &[&str]) -> Output {
    sandbox
        .command()
        .args(arguments)
        .assert()
        .get_output()
        .clone()
}

/// Standard output of a run that had to succeed, quoting stderr when it did not.
fn ok(sandbox: &Sandbox, arguments: &[&str]) -> String {
    let output = run(sandbox, arguments);
    assert_eq!(
        output.status.code(),
        Some(0),
        "`onetaskgraph {}` exited {:?}\n{}",
        arguments.join(" "),
        output.status.code(),
        stderr(&output)
    );
    stdout(&output)
}

/// Standard error of a run that had to fail with `code`.
fn refused(sandbox: &Sandbox, arguments: &[&str], code: i32) -> String {
    let output = run(sandbox, arguments);
    assert_eq!(
        output.status.code(),
        Some(code),
        "`onetaskgraph {}` was expected to exit {code}\n{}{}",
        arguments.join(" "),
        stdout(&output),
        stderr(&output)
    );
    stderr(&output)
}

/// One item of a `--json` copy report, as a comparable triple.
fn reported(rendered: &str) -> Vec<(String, Value, String)> {
    let report: Value = serde_json::from_str(rendered).expect("a copy emits JSON");
    report["items"]
        .as_array()
        .expect("a copy report carries items")
        .iter()
        .map(|item| {
            (
                item["source"].as_str().expect("a source id").to_owned(),
                item["destination"].clone(),
                item["action"].as_str().expect("an action").to_owned(),
            )
        })
        .collect()
}

/// One item of a `<verb> show --json` response.
fn shown(sandbox: &Sandbox, verb: &str, id: &str) -> Value {
    let response: Value =
        serde_json::from_str(&ok(sandbox, &[verb, "show", id, "--json"])).expect("show emits JSON");
    response["items"][0]["item"].clone()
}

/// The persistent peer, configured over `store`.
fn store_at(store: &Path, log: Option<&Path>, documents: &str) -> Value {
    let mut settings = json!({"store": store, "documents": documents});
    if let Some(log) = log {
        settings["log"] = json!(log);
    }
    json!({
        "plugin": "subprocess",
        "config": {
            "command": env!("CARGO_BIN_EXE_onetaskgraph-document-store"),
            "settings": settings,
        },
    })
}

/// The one document this suite copies, with a caller-defined key of every JSON type.
///
/// The types matter as much as the values: a destination that stringified a number, or
/// flattened a nested object, would still round-trip something recognisable, and the whole
/// promise of caller-defined metadata is that it comes back out as it went in.
fn source_document(overrides: Value) -> Value {
    let mut held = json!({
        "id": "D-1",
        "title": "Alpha design",
        "content": "the engine core, reviewed",
        "project": "P-1",
        "labels": [{"id": "L-1", "name": "spec"}],
        "location": {"url": "https://example.invalid/D-1"},
        "metadata": {
            "onepipeline.turn_budget": 12,
            "caller.flags": [true, null],
            "caller.shape": {"nested": {"depth": 2}},
            "caller.ratio": 1.5,
            "caller.note": "a string",
        },
        "repositories": ["github.com/nickderobertis/onetaskgraph"],
    });
    for (key, value) in overrides.as_object().expect("an object of overrides") {
        held[key] = value.clone();
    }
    held
}

/// The configuration these journeys run against: the source, and the persistent store.
fn planted(source: Value, store: &Path) -> String {
    document(&json!({
        SOURCE: {"plugin": "in-memory", "config": {
            "capabilities": {"documents": "native"},
            "projects": [{"id": "P-1", "title": "Engine", "content": null,
                          "status": {"category": "in-progress", "name": "Doing"}, "labels": []}],
            "documents": [source],
        }},
        STORE: store_at(store, None, "native"),
    }))
}

/// Where the destination keeps its documents, under this run's own sandbox.
fn store_path(sandbox: &Sandbox) -> PathBuf {
    sandbox.subdirectory("store").join("documents.json")
}

#[test]
fn a_document_copy_creates_at_a_persistent_destination_and_a_second_copy_updates_that_one() {
    let sandbox = Sandbox::new();
    let store = store_path(&sandbox);
    sandbox.project_document(&planted(source_document(json!({})), &store));

    // One invocation writes.
    let created = reported(&ok(
        &sandbox,
        &[
            "document",
            "copy",
            &qualified(SOURCE, "D-1"),
            "--to",
            STORE,
            "--json",
        ],
    ));
    assert_eq!(
        created,
        vec![(
            qualified(SOURCE, "D-1"),
            json!(qualified(STORE, "D-1")),
            "created".to_owned()
        )]
    );

    // A *different* invocation reads it back — which is the whole point of this file, and
    // which no in-memory destination could have answered.
    let landed = shown(&sandbox, "document", &qualified(STORE, "D-1"));
    assert_eq!(landed["title"], json!("Alpha design"));
    assert_eq!(landed["content"], json!("the engine core, reviewed"));
    assert_eq!(landed["labels"][0]["name"], json!("spec"));
    assert_eq!(landed["project"], json!("P-1"));
    assert_eq!(
        landed["repositories"],
        json!(["github.com/nickderobertis/onetaskgraph"])
    );
    // Every caller-defined key, with its JSON type intact — a number as a number, a list
    // holding two different types, a nested object, a float and a string.
    assert_eq!(landed["metadata"]["onepipeline.turn_budget"], json!(12));
    assert_eq!(landed["metadata"]["caller.flags"], json!([true, null]));
    assert_eq!(
        landed["metadata"]["caller.shape"],
        json!({"nested": {"depth": 2}})
    );
    assert_eq!(landed["metadata"]["caller.ratio"], json!(1.5));
    assert_eq!(landed["metadata"]["caller.note"], json!("a string"));
    // And the origin the copy recorded, which is how the second copy finds this one.
    assert_eq!(
        landed["metadata"]["onetaskgraph.origin"],
        json!(qualified(SOURCE, "D-1"))
    );
    // Where the *source* holds a document says nothing about where the destination does,
    // so the location is the destination's own and this copy did not write one.
    assert_eq!(landed["location"], json!(null));

    // Now the source changes, the way an author editing a document changes it.
    sandbox.project_document(&planted(
        source_document(json!({
            "title": "Alpha design, revised",
            "metadata": {
                "onepipeline.turn_budget": 20,
                "caller.flags": [true, null],
                "caller.shape": {"nested": {"depth": 2}},
                "caller.ratio": 1.5,
                "caller.note": "a string",
            },
        })),
        &store,
    ));

    let updated = reported(&ok(
        &sandbox,
        &[
            "document",
            "copy",
            &qualified(SOURCE, "D-1"),
            "--to",
            STORE,
            "--json",
        ],
    ));
    assert_eq!(
        updated,
        vec![(
            qualified(SOURCE, "D-1"),
            json!(qualified(STORE, "D-1")),
            "updated".to_owned()
        )],
        "a second copy updates the document already there"
    );

    // Exactly one where there was one before, in a later invocation still.
    let listed = ok(&sandbox, &["document", "list", "--source", STORE]);
    assert_eq!(
        listed
            .lines()
            .filter(|line| !line.trim().is_empty())
            .count(),
        1,
        "the second copy updated rather than adding a duplicate:\n{listed}"
    );

    let revised = shown(&sandbox, "document", &qualified(STORE, "D-1"));
    assert_eq!(revised["title"], json!("Alpha design, revised"));
    assert_eq!(revised["metadata"]["onepipeline.turn_budget"], json!(20));
    // Every field the edit did not touch is byte-for-byte what it was.
    for untouched in ["content", "labels", "project", "repositories", "location"] {
        assert_eq!(
            revised[untouched], landed[untouched],
            "the update rewrote {untouched}, which the edit did not touch"
        );
    }
    for untouched in [
        "caller.flags",
        "caller.shape",
        "caller.ratio",
        "caller.note",
        "onetaskgraph.origin",
    ] {
        assert_eq!(
            revised["metadata"][untouched], landed["metadata"][untouched],
            "the update rewrote the metadata key {untouched}, which the edit did not touch"
        );
    }
}

#[test]
fn a_document_copy_into_a_destination_with_no_documents_reads_nothing_from_the_source_first() {
    // The refusal is answered from the declaration the handshake carried, so the proof it
    // owes is not only its own wording: the source has to have been asked *nothing*. This
    // source records every method it is asked for, and afterwards the record holds the
    // handshake and nothing else.
    let sandbox = Sandbox::new();
    let store = store_path(&sandbox);
    let log = sandbox.subdirectory("store").join("asked.log");
    std::fs::write(
        &store,
        json!({"documents": [source_document(json!({}))]}).to_string(),
    )
    .expect("the source's own store");

    sandbox.project_document(&document(&json!({
        SOURCE: store_at(&store, Some(&log), "native"),
        "notes": {"plugin": "local-md", "config": empty_folder(&sandbox, "notes")},
    })));

    let complaint = refused(
        &sandbox,
        &[
            "document",
            "copy",
            &qualified(SOURCE, "D-1"),
            "--to",
            "notes",
        ],
        1,
    );
    assert!(
        complaint.contains("notes")
            && complaint.contains("local-md")
            && complaint.contains("has no documents"),
        "the refusal names the destination and its plugin:\n{complaint}"
    );

    let asked: Vec<String> = std::fs::read_to_string(&log)
        .expect("the source recorded what it was asked")
        .lines()
        .map(str::to_owned)
        .collect();
    assert_eq!(
        asked,
        ["initialize"],
        "the copy was refused before the source was read: it was asked {asked:?}"
    );

    // And nothing was written at the destination either.
    let notes = sandbox.project().join("notes");
    let written: Vec<PathBuf> = walk(&notes);
    assert!(
        written.is_empty(),
        "a refused copy writes nothing: {written:?}"
    );
}

/// This row's own source, with the persistent store beside it as the destination.
fn row_with_store(row: &Row, sandbox: &Sandbox, store: &Path) -> String {
    document(&json!({
        SOURCE: {"plugin": row.plugin, "config": (row.fixture.block)(sandbox)},
        STORE: store_at(store, None, "native"),
    }))
}

#[test]
fn every_document_bearing_row_copies_into_a_persistent_destination_and_is_matched_not_duplicated() {
    // The journey above proves the round trip in full against one source. This proves the
    // half that has to hold for *every* source kind: a document copied out of it lands at a
    // destination that is still there afterwards, reads back as the source reported it, and
    // is matched rather than duplicated when the same copy is run again.
    for row in ROWS
        .iter()
        .filter(|row| row.declared().documents.is_native())
    {
        let sandbox = Sandbox::new();
        let store = store_path(&sandbox);
        sandbox.project_document(&row_with_store(row, &sandbox, &store));

        let created = reported(&ok(
            &sandbox,
            &[
                "document",
                "copy",
                &qualified(SOURCE, "D-1"),
                "--to",
                STORE,
                "--json",
            ],
        ));
        assert_eq!(
            created,
            vec![(
                qualified(SOURCE, "D-1"),
                json!(qualified(STORE, "D-1")),
                "created".to_owned()
            )],
            "{}",
            row.name
        );

        // A later invocation, against the file the first one wrote.
        let landed = shown(&sandbox, "document", &qualified(STORE, "D-1"));
        let source = shown(&sandbox, "document", &qualified(SOURCE, "D-1"));
        for field in ["title", "content", "labels", "project", "repositories"] {
            assert_eq!(
                landed[field], source[field],
                "{}: the destination holds {field} as the source reported it",
                row.name
            );
        }
        for key in ["onepipeline.turn_budget", "caller.flags"] {
            assert_eq!(
                landed["metadata"][key], source["metadata"][key],
                "{}: the destination holds the metadata key {key} with its JSON type intact",
                row.name
            );
        }

        let again = reported(&ok(
            &sandbox,
            &[
                "document",
                "copy",
                &qualified(SOURCE, "D-1"),
                "--to",
                STORE,
                "--json",
            ],
        ));
        assert_eq!(
            again[0].1,
            json!(qualified(STORE, "D-1")),
            "{}: the second copy found the first one's document",
            row.name
        );
        assert!(
            matches!(again[0].2.as_str(), "updated" | "unchanged"),
            "{}: a second copy is not a second create: {:?}",
            row.name,
            again[0]
        );
        let listed = ok(&sandbox, &["document", "list", "--source", STORE]);
        assert_eq!(
            listed
                .lines()
                .filter(|line| !line.trim().is_empty())
                .count(),
            1,
            "{}: exactly one where there was one before:\n{listed}",
            row.name
        );
    }
}

/// Every file under `root`, however deep.
fn walk(root: &Path) -> Vec<PathBuf> {
    let mut found = Vec::new();
    let mut pending = vec![root.to_path_buf()];
    while let Some(directory) = pending.pop() {
        let Ok(entries) = std::fs::read_dir(&directory) else {
            continue;
        };
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                pending.push(path);
            } else {
                found.push(path);
            }
        }
    }
    found
}
